### PR TITLE
Websocket 0.9.2

### PR DESCRIPTION
**Pull request by vbmithr**

### CODE REVIEW

**Comment by avsm** [comment]
The failing test should be addressed by /mirleft/ocaml-tls#253 so merging for now.  Thanks for the fast update Vincent!

File: packages/websocket/websocket.0.9.2/files/websocket.install [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+lib: [
+  "_build/lib/websocket.a" {"websocket.a"}
+  "_build/lib/websocket.cma" {"websocket.cma"}
+  "_build/lib/websocket.cmi" {"websocket.cmi"}
+  "?_build/lib/websocket.cmti" {"websocket.cmti"}
+  "_build/lib/websocket.cmx" {"websocket.cmx"}
+  "_build/lib/websocket.cmxa" {"websocket.cmxa"}
+  "_build/lib/websocket.cmxs" {"websocket.cmxs"}
+  "_build/lib/websocket.mli" {"websocket.mli"}
+  "_build/pkg/META" {"META"} ]
```
Comment (avsm) [inline packages/websocket/websocket.0.9.2/files/websocket.install]: in the future, could you put this file in the repository? It makes pinning easier.  Also, the `cmx/a/s` files can also be optional.

**Comment by vbmithr** [comment]
The `websocket.install` file is auto-generated by `topkg` so it will be there… Is there any circumstances where it will not be generated ?

**Comment by dbuenzli** [comment]
@vbmithr the `topkg` file must be generated on the target host. It should not be in the package metadata.

**Comment by dbuenzli** [comment]
See [here](https://github.com/dbuenzli/topkg#basic-setup) for how `topkg` is supposed to be used.

**Comment by dbuenzli** [comment]
Also [this](https://github.com/vbmithr/ocaml-websocket/blob/master/Makefile#L3) is wrong. You need to pass the correct `opam` variables `%{ocaml-native}%` and `%{ocaml-native-dynlink}%`. As it stands your package will fail to build on byte-code only platforms.

**Comment by vbmithr** [comment]
On 11/02/2015 21:51, Daniel Bünzli wrote:
> Also this
> <https://github.com/vbmithr/ocaml-websocket/blob/master/Makefile#L3> is
> wrong. You need to pass the correct |opam| variables |%{ocaml-native}%|
> and |%{ocaml-native-dynlink}%|. As it stands your package will fail to
> build on byte-code only platforms.
> 

Right, I need to fix this.

However I did not understand your previous messages, or why my usage of
topkg was not the way it should be used. You mean topkg.ml should be
_autogenerated_ in the target platform ? You mention that it is "an
helper to write build.ml files"… I'm a bit lost here.

-- 
Vincent

**Comment by vbmithr** [comment]
On 11/02/2015 21:51, Daniel Bünzli wrote:
> As it stands your package will fail to build on byte-code only platforms.

Ah, yeah, opam and byte-code only platforms… Did you ever try to compile
opam on a byte-code only platform ? :)

I had to do this recently, the result was not pretty pretty to see! :)

-- 
Vincent

**Comment by dbuenzli** [comment]
> You mean topkg.ml should be
> _autogenerated_ in the target platform ?

Sorry no, `pkg/build.ml` must be *invoked* on the target platform (and hence `$PKG.install`generated on the target platform). The message that went into the email was wrong but updated on gh. 

> Did you ever try to compile opam on a byte-code only platform ? :)

Nope but we want to be ready for it. Also native may be available but no dynlink so it's important to correctly thread these variables in the script.

**Comment by vbmithr** [comment]
On 11/02/2015 22:30, Daniel Bünzli wrote:
> Nope but we want to be ready for it. Also native may be available but no
> dynlink so it's important to correctly thread these variables in the
> script.

Fixed. As far as I can tell, I was always invoking pkg/build.ml (through
the Makefile previously, now directly). So I'm still not clear with what
was wrong regarding _not_ including the websocket.install (since it is
autogenerated), or the wrong usage of topkg (I was invoking pkg/build.ml
on the target platform, albeit through a Makefile).

-- 
Vincent

**Comment by dbuenzli** [comment]
> So I'm still not clear with what
> was wrong regarding _not_ including the websocket.install

You [included](https://github.com/ocaml/opam-repository/pull/3553/files#diff-780a51dafcfa5738bddc2ff7235bc8be) it in the opam repository itself that's the problem @avsm mentioned. You should *not* include it in opam respository since it will be generated on the target host (with possibly a different content).  

> (I was invoking pkg/build.ml on the target platform, albeit through a Makefile).

That's fine aswell, however you didn't adapt the `topkg` variables to those of the target host. Now I don't know how well `opam` deals with recursively calling itself to get its variables in that `Makefile`. And one of the advantages of invoking `pkg/build.ml` directly in the `opam` file as mentioned in `topkg`'s documentation rather than trough a makefile is that the invocation gets explicitely output by `opam` which allows to see in a blink how the package is configured (which is useful e.g. when you have `deptopts`, for an example see what `opam (re)install vg` tells you).

**Pull request merged**